File: Flutter/IMSDK/im-flutter-plugin/tencent_im_sdk_plugin_web/lib/src/manager/v2_tim_friendship_manager.dart
// ignore_for_file: avoid_print

import 'dart:js';

import 'package:js/js_util.dart';
import 'package:tencent_im_sdk_plugin_platform_interface/enum/V2TimFriendshipListener.dart';
import 'package:tencent_im_sdk_plugin_platform_interface/models/v2_tim_friend_application.dart';
import 'package:tencent_im_sdk_plugin_platform_interface/models/v2_tim_friend_application_result.dart';
import 'package:tencent_im_sdk_plugin_platform_interface/models/v2_tim_friend_check_result.dart';
import 'package:tencent_im_sdk_plugin_platform_interface/models/v2_tim_friend_group.dart';
import 'package:tencent_im_sdk_plugin_platform_interface/models/v2_tim_friend_info.dart';
import 'package:tencent_im_sdk_plugin_platform_interface/models/v2_tim_friend_info_result.dart';
import 'package:tencent_im_sdk_plugin_platform_interface/models/v2_tim_friend_operation_result.dart';
import 'package:tencent_im_sdk_plugin_web/src/enum/event_enum.dart';
import 'package:tencent_im_sdk_plugin_web/src/enum/friend_type.dart';
import 'package:tencent_im_sdk_plugin_web/src/manager/im_sdk_plugin_js.dart';
import 'package:tencent_im_sdk_plugin_web/src/manager/v2_tim_manager.dart';
import 'package:tencent_im_sdk_plugin_web/src/models/v2_tim_add_friend.dart';
import 'package:tencent_im_sdk_plugin_web/src/models/v2_tim_check_friend.dart';
import 'package:tencent_im_sdk_plugin_web/src/models/v2_tim_friend_black_list.dart';
import 'package:tencent_im_sdk_plugin_web/src/models/v2_tim_friend_group.dart';
import 'package:tencent_im_sdk_plugin_web/src/models/v2_tim_friend_info.dart';
import 'package:tencent_im_sdk_plugin_web/src/models/v2_tim_friend_list.dart';
import 'package:tencent_im_sdk_plugin_web/src/utils/utils.dart';

class V2TIMFriendshipManager {
  late TIM? timeWeb;
  late List<dynamic> friendList;

  V2TIMFriendshipManager() {
    timeWeb = V2TIMManagerWeb.timWeb;
  }

  getUserIDlist(List<dynamic> list) {
    var arr = [];
    for (var item in list) {
      arr.add(item["userID"]);
    }
    return arr;
  }

  getUserIDlistFromJS(List<dynamic> list) {
    var arr = [];
    for (var item in list) {
      arr.add(item.userID);
    }
    return arr;
  }

// 求差集
  getDiffsFromTwoArr(List<dynamic> listA, List<dynamic> listB) {
    var tempArr = [];
    tempArr.addAll(listA);
    tempArr.addAll(listB);
    List<dynamic> difference = tempArr
        .where(
            (element) => !listA.contains(element) || !listB.contains(element))
        .toList();

    return difference;
  }

/*
  这里因为web add Friend 只有一个update
*/
  void setFriendListener(V2TimFriendshipListener listener) {
    // FriendList update linstener
    timeWeb!.on("onFriendListUpdated", allowInterop((res) {
      print("好友列表更新回调执行");
      final formateList =
          FriendList.formatedFriendListRes(jsToMap(res)['data']);
      final oldFriendList = V2TIMManager.getFriendList();
      if (oldFriendList.length == 1 && oldFriendList[0] == "init") {
        return;
      }

      int formateListCount = formateList.length;
      int oldFriendListCount = oldFriendList.length;
      List<dynamic> formateUserIDList = getUserIDlist(formateList);
      List<dynamic> oldFriendUserIDList = getUserIDlist(oldFriendList);

      print("formateListCount:$formateListCount");
      print("oldFriendListCount:$oldFriendListCount");
      if (formateListCount > oldFriendListCount) {
        print("好友列表ADD回调执行");
        final addedList =
            formateList.map((e) => V2TimFriendInfo.fromJson(e)).toList();
        listener.onFriendListAdded(addedList);
      } else {
        print("好友列表Delete回调执行");
        List<dynamic> difference =
            getDiffsFromTwoArr(formateUserIDList, oldFriendUserIDList);
        List<String> userList = List.from(difference);
        listener.onFriendListDeleted(userList);
      }
      V2TIMManager.updtateFriendList(formateList);
    }));

    // BlackList update Listenet
    timeWeb!.on("blacklistUpdated", allowInterop((res) {
      print("黑名单更新回调执行");
      final formateList =
          FriendBlackList.formateBlackList(jsToMap(res)['data']);
      final oldBlackList = V2TIMManager.getBlackList();
      if (oldBlackList.length == 1 && oldBlackList[0] == "init") {
        return;
      }

      int formateListCount = formateList.length;
      int oldBlackListCount = oldBlackList.length;
      List<dynamic> formateUserIDList = getUserIDlist(formateList);
      List<dynamic> oldBlackUserIDList = getUserIDlist(oldBlackList);

      print("formateListCount:$formateListCount");
      print("oldFriendListCount:$oldBlackListCount");
      if (formateListCount > oldBlackListCount) {
        print("好友BlackList Add回调执行");
        final infoListMap =
            formateList.map((e) => V2TimFriendInfo.fromJson(e)).toList();
        listener.onBlackListAdd(infoListMap);
      } else {
        print("好友BlackList Delete回调执行");
        List<dynamic> difference =
            getDiffsFromTwoArr(formateUserIDList, oldBlackUserIDList);
        List<String> userList = List.from(difference);
        listener.onBlackListDeleted(userList);
      }
      V2TIMManager.updateBlackList(formateList);
    }));

    // ApplicationList update Listenet
    timeWeb!.on(EventType.FRIEND_APPLICATION_LIST_UPDATED, allowInterop((res) {
      final resData = jsToMap(res)['data'];
      final List formateList = FriendApplication.formateResult(
          jsToMap(resData))["friendApplicationList"];
      final oldApplicationList = V2TIMManager.getFriendApplicationList();
      if (oldApplicationList.length == 1 && oldApplicationList[0] == "init") {
        return;
      }

      int formateListCount = formateList.length;
      int oldApplicationListCount = oldApplicationList.length;
      List<dynamic> formateUserIDList = getUserIDlist(formateList);
      List<dynamic> oldApplicationIDList = getUserIDlist(oldApplicationList);

      print("formateListCount:$formateListCount");
      print("oldFriendListCount:$oldApplicationListCount");
      if (formateListCount > oldApplicationListCount) {
        print("好友Application ADD回调执行");
        final addedList =
            formateList.map((e) => V2TimFriendApplication.fromJson(e)).toList();
        listener.onFriendApplicationListAdded(addedList);
      } else {
        print("好友Application Delete回调执行");
        List<dynamic> difference =
            getDiffsFromTwoArr(formateUserIDList, oldApplicationIDList);
        List<String> userIDList = List.from(difference);
        listener.onFriendApplicationListDeleted(userIDList);
      }
      V2TIMManager.updateFriendApplicationList(formateList);
    }));

    // FriendIinfo update Listenet, res 只会返回
    timeWeb!.on(EventType.PROFILE_UPDATED, allowInterop((res) async {
      List<dynamic> userIDList = getUserIDlistFromJS(jsToMap(res)['data']);
      print("好友资料更新回调执行");
      // 判断是不是修改自己的信息
      if (userIDList.length == 1 && userIDList[0] == V2TIMManager.getUserID()) {
        return null;
      }

      final result = await wrappedPromiseToFuture(timeWeb!.getFriendProfile(
          FriendInfo.formateGetInfoParams({"userIDList": userIDList})));

      List<dynamic> formateList = FriendInfo.formateFriendInfoToList(
          jsToMap(result.data)['friendList']);
      final infoList =
          formateList.map((e) => V2TimFriendInfo.fromJson(e)).toList();

      listener.onFriendInfoChanged(infoList);
    }));
  }

  void makeConversationListenerEventData(_channel, String type, data) {
    CommonUtils.emitEvent(_channel, "friendListener", type, data);
  }

  Future<dynamic> getFriendList() async {
    try {
      final res = await wrappedPromiseToFuture(timeWeb!.getFriendList());
      final code = res.code;
      final friendList = res.data as List;
      if (code == 0) {
        final formateList = FriendList.formatedFriendListRes(friendList);
        return CommonUtils.returnSuccess<List<V2TimFriendInfo>>(formateList);
      } else {
        return CommonUtils.returnError('获取好友列表失败');
      }
    } catch (error) {
      return CommonUtils.returnError(error);
    }
  }

  Future<dynamic> getFriendsInfo(Map<String, dynamic> params) async {
    try {
      final res = await wrappedPromiseToFuture(
          timeWeb!.getFriendProfile(FriendInfo.formateGetInfoParams(params)));
      final result = await FriendInfo.formateFriendInfoResult(
          jsToMap(res.data)['friendList'],
          jsToMap(res.data)['failureUserIDList']);

      return CommonUtils.returnSuccess<List<V2TimFriendInfoResult>>(result);
    } catch (error) {
      return CommonUtils.returnErrorForValueCb<List<V2TimFriendInfoResult>>(
          error);
    }
  }

  Future<dynamic> addFriend(Map<String, dynamic> params) async {
    try {
      final res = await wrappedPromiseToFuture(
          timeWeb!.addFriend(AddFriend.formateParams(params)));
      final result = AddFriend.formateResult(jsToMap(res.data));

      return CommonUtils.returnSuccess<V2TimFriendOperationResult>(result);
    } catch (error) {
      return CommonUtils.returnErrorForValueCb<V2TimFriendOperationResult>(
          error);
    }
  }

  /*
    web中只有remark和自定义字短可以更新
    /必须以 Tag_SNS_Custom为开头，且此自定义字段需要先到控制台定义
  */
  Future<dynamic> setFriendInfo(Map<String, dynamic> params) async {
    try {
      final formateParams = mapToJSObj({
        "userID": params["userID"],
        "remark": params["friendRemark"],
        "friendCustomField":
            FriendInfo.formateFriendCustomInfo(params["friendCustomInfo"])
      });
      final res =
          await wrappedPromiseToFuture(timeWeb!.updateFriend(formateParams));
      if (res.code == 0) {
        return CommonUtils.returnSuccessForCb(null);
      } else {
        return CommonUtils.returnError(
            "getFriendsInfo faile code:${res.code}, data:${res.data}");
      }
    } catch (error) {
      return CommonUtils.returnError(error);
    }
  }

  Future<dynamic> deleteFromFriendList(Map<String, dynamic> params) async {
    try {
      final formateParams = mapToJSObj({
        "userIDList": params["userIDList"],
        "type": FriendTypeWeb.convertWebFriendType(params["deleteType"]),
      });

      final res =
          await wrappedPromiseToFuture(timeWeb!.deleteFriend(formateParams));

      final successUserIDList = jsToMap(res.data)['successUserIDList'];
      final formateArr = [];
      successUserIDList.forEach((element) =>
          formateArr.add(AddFriend.formateResult(jsToMap(element))));

      return CommonUtils.returnSuccess<List<V2TimFriendOperationResult>>(
          formateArr);
    } catch (error) {
      return CommonUtils.returnErrorForValueCb<
          List<V2TimFriendOperationResult>>(error);
    }
  }

  Future<dynamic> checkFriend(Map<String, dynamic> params) async {
    try {
      final formateParams = CheckFriend.formateParams(params);
      final res =
          await wrappedPromiseToFuture(timeWeb!.checkFriend(formateParams));
      final formateResult = CheckFriend.formateResult(jsToMap(res.data));
      return CommonUtils.returnSuccess<List<V2TimFriendCheckResult>>(
          formateResult);
    } catch (error) {
      return CommonUtils.returnErrorForValueCb<List<V2TimFriendCheckResult>>(
          error);
    }
  }

  Future<dynamic> getFriendApplicationList() async {
    try {
      final res =
          await wrappedPromiseToFuture(timeWeb!.getFriendApplicationList());
      final formateResult = FriendApplication.formateResult(jsToMap(res.data));
      return CommonUtils.returnSuccess<V2TimFriendApplicationResult>(
          formateResult);
    } catch (error) {
      return CommonUtils.returnErrorForValueCb<V2TimFriendApplicationResult>(
          error);
    }
  }

  // 注意：同意好友申请成功无返回值
  Future<dynamic> acceptFriendApplication(Map<String, dynamic> params) async {
    try {
      final formateParams = AcceptFriendApplication.formateParams(params);
      final res = await promiseToFuture(
          timeWeb!.acceptFriendApplication(formateParams));
      if (res != null) {
        log(res);
      }

      return CommonUtils.returnSuccess<V2TimFriendOperationResult>(null);
    } catch (error) {
      return CommonUtils.returnErrorForValueCb<V2TimFriendOperationResult>(
          error);
    }
  }

  // 注意：拒绝好友申请成功无返回值
  Future<dynamic> refuseFriendApplication(Map<String, dynamic> params) async {
    try {
      final formateParams = mapToJSObj({"userID": params["userID"]});
      final result = await promiseToFuture(
          timeWeb!.refuseFriendApplication(formateParams));
      if (result != null) {
        log(result);
      }
      return CommonUtils.returnSuccess<V2TimFriendOperationResult>(null);
    } catch (error) {
      return CommonUtils.returnErrorForValueCb<V2TimFriendOperationResult>(
          error);
    }
  }

  Future<dynamic> deleteFriendApplication(Map<String, dynamic> params) async {
    try {
      final formateParams = mapToJSObj({
        "userID": params["userID"],
        "type": FriendTypeWeb.convertToApplicationFriendType(params["type"])
      });
      final res = await wrappedPromiseToFuture(
          timeWeb!.deleteFriendApplication(formateParams));
      return CommonUtils.returnSuccessForCb(jsToMap(res.data));
    } catch (error) {
      return CommonUtils.returnError(error);
    }
  }

  // 注意：好友申请已读不会返回任何参数
  Future<dynamic> setFriendApplicationRead() async {
    try {
      await promiseToFuture(timeWeb!.setFriendApplicationRead());

      return CommonUtils.returnSuccess(null);
    } catch (error) {
      return CommonUtils.returnError(error);
    }
  }

  Future<dynamic> addToBlackList(params) async {
    try {
      final formateParams = mapToJSObj({
        "userIDList": params["userIDList"],
      });
      final res =
          await wrappedPromiseToFuture(timeWeb!.addToBlacklist(formateParams));
      List<dynamic> resultArr = [];
      res.data.forEach(
          (userID) => resultArr.add({"userID": userID, "resultCode": null}));

      return CommonUtils.returnSuccess<List<V2TimFriendOperationResult>>(
          resultArr);
    } catch (error) {
      return CommonUtils.returnErrorForValueCb<
          List<V2TimFriendOperationResult>>(error);
    }
  }

  Future<dynamic> deleteFromBlackList(params) async {
    try {
      final formateParams = mapToJSObj({
        "userIDList": params["userIDList"],
      });
      final res = await wrappedPromiseToFuture(
          timeWeb!.removeFromBlacklist(formateParams));
      final formateResult = FriendBlackList.formateDeleteBlackListRes(res.data);
      return CommonUtils.returnSuccess<List<V2TimFriendOperationResult>>(
          formateResult);
    } catch (error) {
      return CommonUtils.returnErrorForValueCb<
          List<V2TimFriendOperationResult>>(error);
    }
  }

  // 黑名单这里只会返回userID，无法拿到详细信息
  Future<dynamic> getBlackList() async {
    try {
      final res = await wrappedPromiseToFuture(timeWeb!.getBlacklist());
      final formateResult = FriendBlackList.formateBlackList(res.data);
      log(res);
      return CommonUtils.returnSuccess<List<V2TimFriendInfo>>(formateResult);
    } catch (error) {
      return CommonUtils.returnErrorForValueCb<List<V2TimFriendInfo>>(error);
    }
  }

  Future<dynamic> createFriendGroup(Map<String, dynamic> params) async {
    try {
      var formateParams = FriendGroup.formateParams(params);
      final res = await wrappedPromiseToFuture(
          timeWeb!.createFriendGroup(formateParams));
      log(res);
      final formateResult = FriendGroup.formateResult(jsToMap(res.data));
      return CommonUtils.returnSuccess<List<V2TimFriendOperationResult>>(
          formateResult);
    } catch (error) {
      return CommonUtils.returnErrorForValueCb<
          List<V2TimFriendOperationResult>>(error);
    }
  }

  Future<dynamic> getFriendGroups(Map<String, dynamic> params) async {
    try {
      if (params["groupNameList"] != null) {
        print(" web getFriendGroupList 不支特定分组传入");
      }
      final res = await wrappedPromiseToFuture(timeWeb!.getFriendGroupList());
      final formateResult = FriendGroup.formateGroupResult(res.data);
      return CommonUtils.returnSuccess<List<V2TimFriendGroup>>(formateResult);
    } catch (error) {
      return CommonUtils.returnErrorForValueCb<List<V2TimFriendGroup>>(error);
    }
  }

  // 只支持单项好友分组删除
  Future<dynamic> deleteFriendGroup(Map<String, dynamic> params) async {
    try {
      print(" web deleteFriendGroup 只支持单项好友分组删除");
      final formateParams = mapToJSObj({"name": params["groupNameList"][0]});
      final res =
          await promiseToFuture(timeWeb!.deleteFriendGroup(formateParams));
      log(res);
      return CommonUtils.returnSuccessForCb(null);
    } catch (error) {
      return CommonUtils.returnError(error);
    }
  }

  Future<dynamic> renameFriendGroup(Map<String, dynamic> params) async {
    try {
      final formateParams = mapToJSObj(
          {"oldName": params["oldName"], "newName": params["newName"]});
      final res =
          await promiseToFuture(timeWeb!.renameFriendGroup(formateParams));
      log(res);
      return CommonUtils.returnSuccessForCb(null);
    } catch (error) {
      return CommonUtils.returnError(error);
    }
  }

  Future<dynamic> addFriendsToFriendGroup(Map<String, dynamic> params) async {
    try {
      final formateParams = mapToJSObj(
          {"name": params["groupName"], "userIDList": params["userIDList"]});
      final res = await wrappedPromiseToFuture(
          timeWeb!.addToFriendGroup(formateParams));
      log(res);
      final formateResult = FriendBlackList.formateDeleteBlackListRes(
          jsToMap(jsToMap(res.data)['friendGroup'])['userIDList']);
      return CommonUtils.returnSuccess<List<V2TimFriendOperationResult>>(
          formateResult);
    } catch (error) {
      return CommonUtils.returnErrorForValueCb<
          List<V2TimFriendOperationResult>>(error);
    }
  }

  //分组返回逻辑奇怪 他返回的是现存在分组里的内容，但是natvie中却是删除的
  Future<dynamic> deleteFriendsFromFriendGroup(
      Map<String, dynamic> params) async {
    try {
      final formateParams = mapToJSObj(
          {"name": params["groupName"], "userIDList": params["userIDList"]});
      final res = await wrappedPromiseToFuture(
          timeWeb!.removeFromFriendGroup(formateParams));
      log(res);
      final formateResult = FriendBlackList.formateDeleteBlackListRes(
          jsToMap(jsToMap(res.data)['friendGroup'])['userIDList']);
      return CommonUtils.returnSuccess<List<V2TimFriendOperationResult>>(
          formateResult);
    } catch (error) {
      return CommonUtils.returnErrorForValueCb<
          List<V2TimFriendOperationResult>>(error);
    }
  }

  Future<dynamic> searchFriends() async {
    return CommonUtils.returnErrorForValueCb<List<V2TimFriendInfoResult>>(
        'searchFriends Not support for web');
  }
}

File: ReactNative/IMSDK/react-native-tim-js/src/manager/v2_tim_friendship_manager.ts
/**
 * 腾讯云 IM 在收发消息时默认不检查是不是好友关系，您可以在 [腾讯云IM官网](https://cloud.tencent.com/document/product/269/51940#.E5.A5.BD.E5.8F.8B.E7.AE.A1.E7.90.86.E7.9B.B8.E5.85.B3.E6.8E.A5.E5.8F.A3])
 * 控制台 >功能配置>登录与消息>好友关系检查中开启"发送单聊消息检查关系链"开关，并使用如下接口增删好友和管理好友列表。
 * @module FriendshipManager(好友管理相关接口)
 */
import type { NativeEventEmitter } from 'react-native';
import type V2TimFriendInfo from '../interface/v2TimFriendInfo';
import type V2TimValueCallback from '../interface/v2TimValueCallback';
import type V2TimFriendshipListener from '../interface/v2TimFriendshipListener';
import type V2TimFriendInfoResult from '../interface/v2TimFriendInfoResult';
import type { StringMap } from '../interface/commonInterface';
import type V2TimCallback from '../interface/v2TimCallback';
import type V2TimFriendOperationResult from '../interface/v2TimFriendOperationResult';
import type V2TimFriendCheckResult from '../interface/v2TimFriendCheckResult';
import type V2TimFriendApplicationResult from '../interface/v2TimFriendApplicationResult';
import type V2TimFriendGroup from '../interface/v2TimFriendGroup';
import type V2TimFriendSearchParam from '../interface/v2TimFriendSearchParam';
import type { FriendType } from '../enum/friendType';
import type { FriendResponseTypeEnum } from '../enum/friendResponseType';
import type { FriendApplicationTypeEnum } from '../enum/friendApplicationType';

export class V2TIMFriendshipManager {
    private manager: String = 'friendshipManager';
    private nativeModule: any;
    private friendListenerList: V2TimFriendshipListener[] = [];

    /** @hidden */
    constructor(eventEmitter: NativeEventEmitter, module: any) {
        this.nativeModule = module;
        this.addListener(eventEmitter);
    }

    private callFriendshipListener(eventType: String, data: any) {
        this.friendListenerList.forEach((listener) => {
            switch (eventType) {
                case 'onFriendApplicationListAdded':
                    listener.onFriendApplicationListAdded(data);
                    break;
                case 'onFriendApplicationListDeleted':
                    listener.onFriendApplicationListDeleted(data);
                    break;
                case 'onFriendApplicationListRead':
                    listener.onFriendApplicationListRead();
                    break;
                case 'onFriendListAdded':
                    listener.onFriendListAdded(data);
                    break;
                case 'onFriendListDeleted':
                    listener.onFriendListDeleted(data);
                    break;
                case 'onBlackListAdd':
                    listener.onBlackListAdd(data);
                    break;
                case 'onBlackListDeleted':
                    listener.onBlackListDeleted(data);
                    break;
                case 'onFriendInfoChanged':
                    listener.onFriendInfoChanged(data);
                    break;
            }
        });
    }

    private addListener(eventEmitter: NativeEventEmitter) {
        eventEmitter.addListener('friendListener', (response: any) => {
            const { type, data } = response;
            this.callFriendshipListener(type, data);
        });
    }

    /**
     * 添加关系链监听器
     * @category 事件监听
     * @param listener - 监听器
     */
    public addFriendListener(listener: V2TimFriendshipListener): void {
        if (this.friendListenerList.length == 0) {
            this.nativeModule.call(this.manager, 'addFriendListener', {});
        }
        this.friendListenerList.push(listener);
    }

    /**
     * 移除关系链监听器
     * @category 事件监听
     * @param listener - 监听器
     */
    public removeFriendListener(listener?: V2TimFriendshipListener): void {
        if (!listener) {
            this.friendListenerList = [];
        } else {
            this.friendListenerList = this.friendListenerList.filter(
                (item) => listener != item
            );
        }

        if (this.friendListenerList.length == 0) {
            this.nativeModule.call(this.manager, 'removeFriendListener', {});
        }
    }

    /**
     * 获取好友列表
     * @category 获取好友列表
     */
    public getFriendList(): Promise<V2TimValueCallback<V2TimFriendInfo[]>> {
        return this.nativeModule.call(this.manager, 'getFriendList', {});
    }

    /**
     * 获取指定好友资料
     * @category 获取好友信息
     * @param userIDList - 好友ID数组
     * - ID 建议一次最大 100 个，因为数量过多可能会导致数据包太大被后台拒绝，后台限制数据包最大为 1M。
     */
    public getFriendsInfo(
        userIDList: String[]
    ): Promise<V2TimValueCallback<V2TimFriendInfoResult[]>> {
        return this.nativeModule.call(this.manager, 'getFriendsInfo', {
            userIDList,
        });
    }

    /**
     * 设置指定好友资料
     * @category 设置指定好友资料
     * @param userID - 好友userID
     * @param friendRemark - 好友备注
     * @param friendCustomInfo - 好友自定义信息
     */
    public setFriendInfo(
        userID: String,
        friendRemark?: String,
        friendCustomInfo?: StringMap
    ): Promise<V2TimCallback> {
        return this.nativeModule.call(this.manager, 'setFriendInfo', {
            userID,
            friendRemark,
            friendCustomInfo,
        });
    }

    /**
     * ### 添加好友
     * @param userID - 好友userID
     * @param addType - 添加方式
     * @param remark - 好友备注
     * @param friendGroup - 好友分组
     * @param addWording - 添加简述
     * @param addSource - 来源描述
     */
    public addFriend(
        userID: String,
        addType: FriendType,
        remark?: String,
        friendGroup?: String,
        addWording?: String,
        addSource?: String
    ): Promise<V2TimValueCallback<V2TimFriendOperationResult>> {
        return this.nativeModule.call(this.manager, 'addFriend', {
            userID,
            addType,
            remark,
            friendGroup,
            addWording,
            addSource,
        });
    }

    /**
     * 删除好友
     * @param userIDList 要删除的好友 userID 列表
     * @param deleteType 删除类型
     */
    public deleteFromFriendList(
        userIDList: String[],
        deleteType: FriendType
    ): Promise<V2TimValueCallback<V2TimFriendOperationResult[]>> {
        return this.nativeModule.call(this.manager, 'deleteFromFriendList', {
            userIDList,
            deleteType,
        });
    }

    /**
     * 检查指定用户的好友关系
     * @param userIDList 要检查的userID 列表
     * @param checkType 检查类型
     *
     * @note
     * checkType 的使用需要注意：
     * - checkType 如果传入 V2TIM_FRIEND_TYPE_SINGLE，结果返回：V2TIM_FRIEND_RELATION_TYPE_NONE、V2TIM_FRIEND_RELATION_TYPE_IN_MY_FRIEND_LIST 两种情况
     * - checkType 如果传入 V2TIM_FRIEND_TYPE_BOTH，结果返回：V2TIM_FRIEND_RELATION_TYPE_NONE、V2TIM_FRIEND_RELATION_TYPE_IN_MY_FRIEND_LIST、 V2TIM_FRIEND_RELATION_TYPE_IN_OTHER_FRIEND_LIST、V2TIM_FRIEND_RELATION_TYPE_BOTH_WAY 四种情况
     */
    public checkFriend(
        userIDList: String[],
        checkType: FriendType
    ): Promise<V2TimValueCallback<V2TimFriendCheckResult[]>> {
        return this.nativeModule.call(this.manager, 'checkFriend', {
            userIDList,
            checkType,
        });
    }

    /**
     * 获取好友申请列表
     */
    public getFriendApplicationList(): Promise<
        V2TimValueCallback<V2TimFriendApplicationResult>
    > {
        return this.nativeModule.call(
            this.manager,
            'getFriendApplicationList',
            {}
        );
    }

    /**
     * ### 同意好友申请
     * @param responseType - 建立单向/双向好友关系
     * @param type - 请求类型
     * @param userID - 用户ID
     */
    public acceptFriendApplication(
        responseType: FriendResponseTypeEnum,
        type: FriendApplicationTypeEnum,
        userID: String
    ): Promise<V2TimValueCallback<V2TimFriendOperationResult>> {
        return this.nativeModule.call(this.manager, 'acceptFriendApplication', {
            responseType,
            type,
            userID,
        });
    }

    /**
     * ### 拒绝好友申请
     * @param type 请求类型
     * @param userID 用户ID
     */
    public refuseFriendApplication(
        type: FriendApplicationTypeEnum,
        userID: String
    ): Promise<V2TimValueCallback<V2TimFriendOperationResult>> {
        return this.nativeModule.call(this.manager, 'refuseFriendApplication', {
            type,
            userID,
        });
    }

    /**
     * ### 删除好友申请
     * @param type 请求类型
     * @param userID 用户ID
     */
    public deleteFriendApplication(
        type: FriendApplicationTypeEnum,
        userID: String
    ): Promise<V2TimValueCallback<V2TimFriendOperationResult>> {
        return this.nativeModule.call(this.manager, 'deleteFriendApplication', {
            type,
            userID,
        });
    }

    /**
     * ### 设置好友申请已读
     */
    public setFriendApplicationRead(): Promise<V2TimCallback> {
        return this.nativeModule.call(
            this.manager,
            'setFriendApplicationRead',
            {}
        );
    }

    /**
     * 添加用户到黑名单
     * @param userIDList 用户userID 列表
     */
    public addToBlackList(
        userIDList: String[]
    ): Promise<V2TimValueCallback<V2TimFriendOperationResult[]>> {
        return this.nativeModule.call(this.manager, 'addToBlackList', {
            userIDList,
        });
    }

    /**
     * 把用户从黑名单中删除
     * @param userIDList 用户userID 列表
     */
    public deleteFromBlackList(
        userIDList: String[]
    ): Promise<V2TimValueCallback<V2TimFriendOperationResult[]>> {
        return this.nativeModule.call(this.manager, 'deleteFromBlackList', {
            userIDList,
        });
    }

    /**
     * ### 获取黑名单
     */
    public getBlackList(): Promise<V2TimValueCallback<V2TimFriendInfo[]>> {
        return this.nativeModule.call(this.manager, 'getBlackList', {});
    }

    /**
     * ### 创建好友分组
     * @param groupName - 分组名称
     * @param userIDList - 用户userID 列表
     */
    public createFriendGroup(
        groupName: String,
        userIDList?: String[]
    ): Promise<V2TimValueCallback<V2TimFriendOperationResult[]>> {
        return this.nativeModule.call(this.manager, 'createFriendGroup', {
            groupName,
            userIDList,
        });
    }

    /**
     * ### 获取分组信息
     * @param groupNameList - 要获取信息的好友分组名称列表,传入 null 获得所有分组信息
     */
    public getFriendGroups(
        groupNameList?: String[]
    ): Promise<V2TimValueCallback<V2TimFriendGroup[]>> {
        return this.nativeModule.call(this.manager, 'getFriendGroups', {
            groupNameList,
        });
    }

    /**
     * ### 删除好友分组
     * @param groupNameList - 分组名称列表
     */
    public deleteFriendGroup(groupNameList?: String[]): Promise<V2TimCallback> {
        return this.nativeModule.call(this.manager, 'deleteFriendGroup', {
            groupNameList,
        });
    }

    /**
     * ### 好友分组重命名
     * @param oldName - 分组旧名称
     * @param newName - 分组新名称
     */
    public renameFriendGroup(
        oldName: String,
        newName: String
    ): Promise<V2TimCallback> {
        return this.nativeModule.call(this.manager, 'renameFriendGroup', {
            oldName,
            newName,
        });
    }

    /**
     * ### 添加好友到分组
     * @param groupName - 好友分组名称
     * @param userIDList - 用户userID 列表
     */
    public addFriendsToFriendGroup(
        groupName: String,
        userIDList: String[]
    ): Promise<V2TimValueCallback<V2TimFriendOperationResult[]>> {
        return this.nativeModule.call(this.manager, 'addFriendsToFriendGroup', {
            groupName,
            userIDList,
        });
    }

    /**
     * ### 把好友从好友分组中删除
     * @param groupName - 好友分组名称
     * @param userIDList - 用户userID 列表
     */
    public deleteFriendsFromFriendGroup(
        groupName: String,
        userIDList: String[]
    ): Promise<V2TimValueCallback<V2TimFriendOperationResult[]>> {
        return this.nativeModule.call(
            this.manager,
            'deleteFriendsFromFriendGroup',
            { groupName, userIDList }
        );
    }

    /**
     * ### 搜索好友
     * @param param0
     * @param param0.keywordList - 搜索关键字列表，关键字列表最多支持5个。
     * @param param0.isSearchNickName - 是否搜索昵称
     * @param param0.isSearchRemark - 是否搜索备注
     * @param param0.isSearchUserID - 是否搜索 userID
     *
     * @note
     * 需要购买旗舰版套餐
     */
    public searchFriends({
        keywordList,
        isSearchNickName = true,
        isSearchRemark = true,
        isSearchUserID = true,
    }: V2TimFriendSearchParam): Promise<
        V2TimValueCallback<V2TimFriendInfoResult[]>
    > {
        return this.nativeModule.call(this.manager, 'searchFriends', {
            searchParam: {
                keywordList,
                isSearchRemark,
                isSearchNickName,
                isSearchUserID,
            },
        });
    }
}
